### PR TITLE
Calculate query peak memory usage as memory cost

### PR DESCRIPTION
**Pull request by wenleix**
Extracted-From: https://github.com/prestosql/presto/pull/247

### CODE REVIEW

**Comment by wenleix** [comment]
hmm.. guess some error happened during cherry-pick so there is compilation failures...

File: presto-main/src/main/java/com/facebook/presto/cost/CostCalculatorUsingExchanges.java [hunk 1]
```diff
@@ -259,40 +251,40 @@ public PlanCostEstimate visitSpatialJoin(SpatialJoinNode node, Void context)
...
             // Cost will be accounted either in CostCalculatorUsingExchanges#CostEstimator#visitExchanged
             // or in CostCalculatorWithEstimatedExchanges#CostEstimator#visitUnion
             // This stub is needed just to avoid the cumulative cost being set to unknown
-            return PlanCostEstimate.zero();
+            return LocalCostEstimate.zero();
         }
 
         @Override
-        public PlanCostEstimate visitSort(SortNode node, Void context)
+        public LocalCostEstimate visitSort(SortNode node, Void context)
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/cost/CostCalculatorUsingExchanges.java]: I add this change during backport. Looks like when @findepi ports https://github.com/prestodb/presto/pull/12347 via https://github.com/prestosql/presto/pull/336, `visitSort` is not ported in PrestoSQL's version :) 

cc @picrinite , @rschlussel , @arhimondr

File: presto-main/src/main/java/com/facebook/presto/cost/CostCalculatorUsingExchanges.java [hunk 2]
```diff
@@ -244,73 +250,133 @@ public LocalCostEstimate visitExchange(ExchangeNode node, Void context)
...
+                    max(
+                            probeCost.getMaxMemory() + buildCost.getMaxMemory(), // Probe and build execute independently, so their max memory allocations can be realized at the same time
+                            probeCost.getMaxMemory() + buildCost.getMaxMemoryWhenOutputting() + localCost.getMaxMemory()),
+                    probeCost.getMaxMemoryWhenOutputting() + localCost.getMaxMemory(), // Build side finished and freed its memory allocations
+                    probeCost.getNetworkCost() + buildCost.getNetworkCost() + localCost.getNetworkCost());
         }
 
         @Override
-        public LocalCostEstimate visitSort(SortNode node, Void context)
+        public PlanCostEstimate visitSort(SortNode node, Void context)
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/cost/CostCalculatorUsingExchanges.java]: Also changed here (in the last commit)

File: presto-main/src/main/java/com/facebook/presto/cost/CostCalculatorUsingExchanges.java [hunk 3]
```diff
@@ -244,73 +250,133 @@ public LocalCostEstimate visitExchange(ExchangeNode node, Void context)
...
         @Override
-        public LocalCostEstimate visitLimit(LimitNode node, Void context)
+        public PlanCostEstimate visitLimit(LimitNode node, Void context)
         {
             // This is just a wild guess. First of all, LimitNode is rather rare except as a top node of a query plan,
             // so proper cost estimation is not that important. Second, since LimitNode can lead to incomplete evaluation
             // of the source, true cost estimation should be implemented as a "constraint" enforced on a sub-tree and
             // evaluated in context of actual source node type (and their sources).
-            return LocalCostEstimate.ofCpu(getStats(node).getOutputSizeInBytes(node.getOutputSymbols(), types));
+            LocalCostEstimate localCost = LocalCostEstimate.ofCpu(getStats(node).getOutputSizeInBytes(node.getOutputSymbols(), types));
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/cost/CostCalculatorUsingExchanges.java]: @findepi Just curious, why does the Limit operation has any CPU cost?

File: presto-main/src/main/java/com/facebook/presto/cost/CostCalculatorUsingExchanges.java (hunk 2 above)
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/cost/CostCalculatorUsingExchanges.java]: Move it after `visitUnion`

File: presto-main/src/main/java/com/facebook/presto/cost/CostCalculatorUsingExchanges.java [hunk 4]
```diff
@@ -244,73 +250,133 @@ public LocalCostEstimate visitExchange(ExchangeNode node, Void context)
...
+                    probeCost.getMaxMemoryWhenOutputting() + localCost.getMaxMemory(), // Build side finished and freed its memory allocations
+                    probeCost.getNetworkCost() + buildCost.getNetworkCost() + localCost.getNetworkCost());
         }
 
         @Override
-        public LocalCostEstimate visitSort(SortNode node, Void context)
+        public PlanCostEstimate visitSort(SortNode node, Void context)
         {
-            return LocalCostEstimate.ofCpu(getStats(node).getOutputSizeInBytes(node.getOutputSymbols(), types));
+            LocalCostEstimate localCost = LocalCostEstimate.ofCpu(getStats(node).getOutputSizeInBytes(node.getOutputSymbols(), types));
```
[the remark this answers is not in the document]
Comment (wenleix) [reply presto-main/src/main/java/com/facebook/presto/cost/CostCalculatorUsingExchanges.java]: hmm... Now I think about it, maybe the Cpu cost should also multiply by some kind of `log (input size)` ?

[the remark this answers is not in the document]
Comment (wenleix) [reply presto-main/src/main/java/com/facebook/presto/cost/CostCalculatorUsingExchanges.java]: anyhow, that's the cpu cost behavior introduced before :). I will not change it in this PR .

**Comment by wenleix** [comment]
This can be tangentially  related to https://github.com/prestodb/presto/issues/12387 once we want to decide materializing exchange based on memory cost.

**Pull request merged**